src/main/kotlin/com/demonwav/mcdev/i18n/translations/TranslationFunction.kt

```diff
@@ -0,0 +1,134 @@
+                    return Step(true, single, result)
+                }
+                return Step(false, true, if (acc.key.contains(I18nReference.VARIABLE_MARKER)) result.replace(I18nReference.VARIABLE_MARKER, acc.key) else result)
+            }
+            return null
+        }
+
+        val calls = getCalls(call, matchedIndex)
+        val referenced = getMethod(call) ?: return null
+        val result = calls.foldIndexed(Step(false, true, "") as Step?,
+            { depth, acc, v ->
+                if (acc == null) acc else resolveCall(depth, calls.count() == 1, referenced, v, acc)
+            })?.result
+        return result?.copy(second = prefix + result.second + suffix)
+    }
+
+    fun format(translation: String, topCall: PsiCall): Pair<String, Int>? {
+        if (!formatting) {
+            return translation to -1
+        }
+        val format = NUMBER_FORMATTING_PATTERN.matcher(translation).replaceAll("%$1s")
+
+        fun resolveCall(call: PsiCall, substitutions: Map<Int, Array<String?>?>): Map<Int, Array<String?>?> {
+            val method = call.referencedMethod
+            val args = call.argumentList?.expressions
+            return if (method != null && args != null && args.size >= method.parameterList.parametersCount)
+                method.parameterList.parameters
+                    .mapIndexed { i, parameter ->
+                        if (parameter.isVarArgs) {
+                            val varargType = method.getSignature(PsiSubstitutor.EMPTY).parameterTypes[i]
+                            Pair(i, extractVarArgs(varargType, args.drop(i), substitutions, true, true))
+                        } else {
+                            Pair(i, args[i].substituteParameter(substitutions, true, true))
+                        }
+                    }.toMap()
+            else emptyMap()
+        }
+
+        val calls = getCalls(topCall, matchedIndex)
+        val paramCount = STRING_FORMATTING_PATTERN.matcher(format).let {
+            var count = 0
+            while (it.find()) {
+                count++
+            }
+            return@let count
+        }
+        val substitutions = if (calls.count() > 1) calls.take(calls.count() - 1).fold(emptyMap<Int, Array<String?>?>(), { acc, v -> resolveCall(v, acc) }) else emptyMap()
+        val referenceCall = if (calls.count() > 1) calls.last() else calls.first()
+        val method = referenceCall.referencedMethod ?: return translation to -1
+        val varargs = referenceCall.extractVarArgs(method.parameterList.parametersCount - 1, substitutions, calls.count() <= 1, true)
+        val varargStart = if (varargs.size > paramCount) method.parameterList.parametersCount - 1 + paramCount else -1
+        return String.format(format, *varargs) to varargStart
+    }
+
+    override fun toString(): String {
+        return "$memberReference@$matchedIndex"
+    }
+
+    companion object {
+        val NUMBER_FORMATTING_PATTERN = Pattern.compile("%(\\d+\\$)?[\\d.]*[df]")
```

Comment:
These either need the type explicitly defined (since it's inferred from a Java message), or use Kotlin's `Regex` instead.